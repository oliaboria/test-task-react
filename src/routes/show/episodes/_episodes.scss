@import '../../../assets/style/_color';
@import '../../../assets/style/_typography';

.episodes {
    &-title {
        @include h3-heading;
        margin-bottom: 10px;
        color: $green;
    }
}
