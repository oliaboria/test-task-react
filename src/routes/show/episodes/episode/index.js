export { default } from './episode.component';
