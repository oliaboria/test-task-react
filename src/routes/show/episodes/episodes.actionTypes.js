export const LOAD_EPISODES = 'show@LOAD_EPISODES';
export const LOAD_EPISODES_SUCCESS = 'show@LOAD_EPISODES_SUCCESS';
export const LOAD_EPISODES_FAIL = 'show@LOAD_EPISODES_FAIL';
