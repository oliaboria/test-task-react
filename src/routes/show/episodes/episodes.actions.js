import { fetchEpisodeList } from '../../../api';

import {
    LOAD_EPISODES,
    LOAD_EPISODES_SUCCESS,
    LOAD_EPISODES_FAIL,
} from './episodes.actionTypes';

export const loadEpisodesAction = (showId) => ({
    type: LOAD_EPISODES,
    payload: { showId, isLoading: true, isError: false },
});

export const loadEpisodesSuccessAction = (episodes) => ({
    type: LOAD_EPISODES_SUCCESS,
    payload: { episodes, isLoading: false },
});

export const loadEpisodesFailAction = (error) => ({
    type: LOAD_EPISODES_FAIL,
    payload: { error, isLoading: false, isError: true },
});

const getEpisodesByShowId = (showId) => async (dispatch) => {
    try {
        dispatch(loadEpisodesAction(showId));

        const episodes = await fetchEpisodeList(showId);

        dispatch(loadEpisodesSuccessAction(episodes));
    } catch (err) {
        dispatch(loadEpisodesFailAction(err));
    }
};

export default getEpisodesByShowId;
