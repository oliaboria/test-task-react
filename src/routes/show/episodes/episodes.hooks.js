import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

import getEpisodesByShowId from './episodes.actions';
import {
    selectIsEpisodesLoading,
    selectEpisodes,
    selectNumberOfSeasons,
} from './episodes.selectors';

const useAsyncEpisodes = () => {
    const { showId } = useParams();
    const dispatch = useDispatch();

    const episodes = useSelector(selectEpisodes);
    const seasonsNumber = useSelector(selectNumberOfSeasons);
    const isEpisodesLoading = useSelector(selectIsEpisodesLoading);

    useEffect(() => {
        dispatch(getEpisodesByShowId(showId));
    }, [showId, dispatch]);

    return { episodes, seasonsNumber, isEpisodesLoading };
};

export default useAsyncEpisodes;
