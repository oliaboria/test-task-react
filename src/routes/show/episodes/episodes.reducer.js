import groupBy from 'lodash.groupby';

import {
    LOAD_EPISODES,
    LOAD_EPISODES_SUCCESS,
    LOAD_EPISODES_FAIL,
} from './episodes.actionTypes';

const initialState = {
    showId: null,
    episodesBySeason: {},
    seasonsNumber: 0,
    isLoading: false,
    isError: false,
};

function episodesReducer(state = initialState, action) {
    switch (action.type) {
        case LOAD_EPISODES:
        case LOAD_EPISODES_FAIL:
            return { ...state, ...action.payload };
        case LOAD_EPISODES_SUCCESS: {
            const { episodes, isLoading } = action.payload;
            const episodesBySeason = groupBy(episodes, 'season');
            const seasonsNumber = Object.keys(episodesBySeason).length;

            return { ...state, episodesBySeason, seasonsNumber, isLoading };
        }
        default:
            return state;
    }
}

export default episodesReducer;
