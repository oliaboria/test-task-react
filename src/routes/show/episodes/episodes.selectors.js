export const selectEpisodesState = (state) => state.episodes;

export const selectEpisodes = (state) =>
    selectEpisodesState(state).episodesBySeason;

export const selectNumberOfSeasons = (state) =>
    selectEpisodesState(state).seasonsNumber;

export const selectIsEpisodesLoading = (state) =>
    selectEpisodesState(state).isLoading;
