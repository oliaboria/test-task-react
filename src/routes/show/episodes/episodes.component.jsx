import { shape, number, arrayOf } from 'prop-types';
import React from 'react';

import './_episodes.scss';

import episodeType from '../../../types/episode.type';

import EpisodeList from './episodeList';

const Episodes = ({ episodes, seasonsNumber, showId }) => {
    return (
        <>
            {Array(seasonsNumber)
                .fill(seasonsNumber)
                .map((_, index) => {
                    const numberOfSeason = index + 1;

                    return (
                        <section
                            key={`season-${numberOfSeason}`}
                            className="episodes">
                            <h3 className="episodes-title">
                                Season {numberOfSeason}
                            </h3>
                            <EpisodeList
                                episodes={episodes[numberOfSeason]}
                                showId={showId}
                            />
                        </section>
                    );
                })}
        </>
    );
};

Episodes.propTypes = {
    episodes: shape({
        [number]: arrayOf(episodeType),
    }).isRequired,
    seasonsNumber: number.isRequired,
    showId: number.isRequired,
};

Episodes.defaultProps = {
    episodes: {},
    showId: null,
    seasonsNumber: null,
};

export default Episodes;
