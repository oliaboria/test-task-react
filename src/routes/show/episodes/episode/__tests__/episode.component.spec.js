import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import { useParams } from 'react-router-dom';

import { mockEpisodes } from '../../../../../mocks/episodes';
import Episode from '../episode.component';

jest.mock('react-router-dom', () => ({
    useParams: jest.fn(),
    Link: () => <mock-link />,
}));

describe('Episode component', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render episode', () => {
        useParams.mockImplementation(() => ({
            showId: 111,
        }));

        const wrapper = shallow(<Episode episode={mockEpisodes[0]} />);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
});
