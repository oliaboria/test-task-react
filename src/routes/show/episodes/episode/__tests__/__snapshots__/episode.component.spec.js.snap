// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episode component should render episode 1`] = `
<tr>
  <td>
    1
  </td>
  <td>
    Apr 04, 2016
  </td>
  <td>
    <Link
      className="episode-link"
      to={
        Object {
          "pathname": "/shows/111/episodes/657308",
          "search": "?season=1&number=1",
        }
      }
    >
      Escape from Monster Island
    </Link>
  </td>
</tr>
`;
